## 수정사항
#### 2021.06.06
* T_ej는 마지막 조인트 각도에 영향을 받음, 하지만 Q는 unkown이므로 특정하는게 불가능, T_ee로 전부 새로 학습할 것.

In [1]:
from tensorflow.compat.v1 import ConfigProto
from tensorflow.compat.v1 import InteractiveSession

config = ConfigProto()
config.gpu_options.allow_growth = True
session = InteractiveSession(config=config)

In [2]:
import os
import shutil
import random
PROJ_DIR = os.environ["RNB_PLANNING_DIR"]
os.chdir(os.path.join(PROJ_DIR, "src"))

from pkg.utils.utils_python3 import *
from pkg.planning.filtering.lattice_model.data_utils import *

DATA_PATH = os.path.join(PROJ_DIR, "data")
LAT_DATA_PATH = os.path.join(DATA_PATH, "latticized")
MODEL_PATH = os.path.join(PROJ_DIR, "model")
LAT_MODEL_PATH = os.path.join(MODEL_PATH,"latticized")
try_mkdir(MODEL_PATH)
try_mkdir(LAT_MODEL_PATH)
GRASP_FOLDER = "grasp"
ARM10_FOLDER = "arm_10"
ARM05_FOLDER = "arm_05"
FULLS_FOLDER = "full_scene"

In [3]:
ROBOT_TYPE_NAME="panda"
ROBOT_DATA_ROOT = os.path.join(LAT_DATA_PATH, ROBOT_TYPE_NAME)
ROBOT_DATA_ROOT_FAILMORE = ROBOT_DATA_ROOT+"-failmore"
ROBOT_MODEL_ROOT =  os.path.join(LAT_MODEL_PATH, ROBOT_TYPE_NAME)
ARM_FOLDER = ARM10_FOLDER

In [4]:
dataset_list = sorted(os.listdir(ROBOT_DATA_ROOT))
DATASET_TRAIN_FILTERED = dataset_list[:10]
DATASET_TEST_FILTERED = dataset_list[10:15]
print(DATASET_TRAIN_FILTERED)
print(DATASET_TEST_FILTERED)

dataset_list_fails = sorted([folder for folder 
                             in os.listdir(ROBOT_DATA_ROOT_FAILMORE) 
                             if not folder.startswith(".")])
DATASET_TRAIN_FAILMORE = dataset_list_fails[:10]
DATASET_TEST_FAILMORE = dataset_list_fails[10:15]
print(DATASET_TRAIN_FAILMORE)
print(DATASET_TEST_FAILMORE)

['20210920-194311', '20210920-233941', '20210921-032259', '20210921-072543', '20210921-111615', '20210921-150638', '20210921-192001', '20210921-232015', '20210922-033134', '20210922-074138']
['20210922-113744', '20210922-153855', '20210922-193840', '20210922-233844', '20210923-040005']
['20210923-084301', '20210923-090916', '20210923-092756', '20210923-094543', '20210923-101344', '20210923-103627', '20210923-110151', '20210923-112256', '20210923-114722', '20210923-121506']
['20210923-123816', '20210923-130206', '20210923-132926', '20210923-135036', '20210923-141543']


In [5]:
data_pairs_train_filtered = []
for dataset in DATASET_TRAIN_FILTERED:
    file_list = sorted(os.listdir(os.path.join(ROBOT_DATA_ROOT, dataset, GRASP_FOLDER)))
    for file in file_list:
        data_pairs_train_filtered.append((os.path.join(ROBOT_DATA_ROOT, dataset, GRASP_FOLDER, file), 
                                 os.path.join(ROBOT_DATA_ROOT, dataset, ARM_FOLDER, file)))
print("train set: {}".format(len(data_pairs_train_filtered)))        
        
        
data_pairs_test_filtered = []
for dataset in DATASET_TEST_FILTERED:
    file_list = sorted(os.listdir(os.path.join(ROBOT_DATA_ROOT, dataset, GRASP_FOLDER)))
    for file in file_list:
        data_pairs_test_filtered.append((os.path.join(ROBOT_DATA_ROOT, dataset, GRASP_FOLDER, file), 
                                 os.path.join(ROBOT_DATA_ROOT, dataset, ARM_FOLDER, file)))
print("test set: {}".format(len(data_pairs_test_filtered)))        

data_pairs_train_failmore = []
for dataset in DATASET_TRAIN_FAILMORE:
    file_list = sorted(os.listdir(os.path.join(ROBOT_DATA_ROOT_FAILMORE, dataset, GRASP_FOLDER)))
    for file in file_list:
        data_pairs_train_failmore.append((os.path.join(ROBOT_DATA_ROOT_FAILMORE, dataset, GRASP_FOLDER, file), 
                                 os.path.join(ROBOT_DATA_ROOT_FAILMORE, dataset, ARM_FOLDER, file)))
print("train fail set: {}".format(len(data_pairs_train_failmore)))        
        
        
data_pairs_test_failmore = []
for dataset in DATASET_TEST_FAILMORE:
    file_list = sorted(os.listdir(os.path.join(ROBOT_DATA_ROOT_FAILMORE, dataset, GRASP_FOLDER)))
    for file in file_list:
        data_pairs_test_failmore.append((os.path.join(ROBOT_DATA_ROOT_FAILMORE, dataset, GRASP_FOLDER, file), 
                                 os.path.join(ROBOT_DATA_ROOT_FAILMORE, dataset, ARM_FOLDER, file)))
print("test fail set: {}".format(len(data_pairs_test_failmore)))        

train set: 10003
test set: 5000
train fail set: 10038
test fail set: 5022


In [6]:
N_max_train = 10000
N_max_test = 5000

print("=== mix data ===")
data_pairs_train_mixed = mix_data_pairs(
    data_pairs_train_filtered, data_pairs_train_failmore, N_max_train)
data_pairs_test_mixed = mix_data_pairs(
    data_pairs_test_filtered, data_pairs_test_failmore, N_max_test)

data_pairs_train_filtered = data_pairs_train_filtered[:N_max_train]
data_pairs_test_filtered = data_pairs_test_filtered[:N_max_test]
data_pairs_train_failmore = data_pairs_train_failmore[:N_max_train]
data_pairs_test_failmore = data_pairs_test_failmore[:N_max_test]

print("="*20)
print("filtered data: {} / {}".format(
    len(data_pairs_train_filtered), 
    len(data_pairs_test_filtered)))
print("failmore data: {} / {}".format(
    len(data_pairs_train_failmore), 
    len(data_pairs_test_failmore)))
print("mixed data: {} / {}".format(
    len(data_pairs_train_mixed), 
    len(data_pairs_test_mixed)))

=== mix data ===
filtered success: 5000
success from failmore: 0
dataset fails: 5000
dataset all: 10000
filtered success: 2500
success from failmore: 0
dataset fails: 2500
dataset all: 5000
filtered data: 10000 / 5000
failmore data: 10000 / 5000
mixed data: 10000 / 5000


## Train

In [7]:
from pkg.planning.filtering.lattice_model.lattice_model import *

# Create an instance of the model
model = ResNetModelTP()

loss_object = tf.keras.losses.BinaryCrossentropy()
optimizer = tf.keras.optimizers.Adam()

train_loss = tf.keras.metrics.Mean(name='train_loss')
train_accuracy = tf.keras.metrics.BinaryAccuracy(name='train_accuracy')
@tf.function
def train_step(images, labels):
    with tf.GradientTape() as tape:
        # training=True is only needed if there are layers with different
        # behavior during training versus inference (e.g. Dropout).
        predictions = model(images, training=True)
        loss = loss_object(labels, predictions)
    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))

    train_loss(loss)
    train_accuracy(labels, predictions)
    

test_loss = tf.keras.metrics.Mean(name='test_loss')
test_accuracy = tf.keras.metrics.BinaryAccuracy(name='test_accuracy')
@tf.function
def test_step(images, labels):
    # training=False is only needed if there are layers with different
    # behavior during training versus inference (e.g. Dropout).
    predictions = model(images, training=False)
    t_loss = loss_object(labels, predictions)

    test_loss(t_loss)
    test_accuracy(labels, predictions)

    
def train_script(data_pairs_train, data_pairs_test, robot_model_root,
                 EPOCHS_S=0, EPOCHS_E=15, BATCH_SIZE=16, LOG_STEP=100):
    current_time = get_now()
    logpath = os.path.join(robot_model_root, current_time)
    try_mkdir(logpath)
    train_log_dir = os.path.join(logpath, 'train')
    test_log_dir = os.path.join(logpath, 'test')
    model_log_dir = os.path.join(logpath, 'model_{}/')
    train_summary_writer = tf.summary.create_file_writer(train_log_dir)
    test_summary_writer = tf.summary.create_file_writer(test_log_dir)
    shutil.copy(os.path.join(
        PROJ_DIR,'src', 'pkg','planning','filtering','lattice_model','lattice_model.py' ), logpath)
    print(f'Log path: {logpath}')
    
    N_train = len(data_pairs_train)
    N_test = len(data_pairs_test)
    random.shuffle(data_pairs_test)
    res_list = []
    
    for epoch in range(EPOCHS_S, EPOCHS_E):
        # Reset the metrics at the start of the next epoch
        train_loss.reset_states()
        train_accuracy.reset_states()
        test_loss.reset_states()
        test_accuracy.reset_states()

        random.shuffle(data_pairs_train)
        i_step = 0
        data_batch, label_batch = [], []
        for data_pair in data_pairs_train:
            i_step += 1
            grasp_img, arm_img, rh_mask, label = load_data(data_pair)
            data_batch.append([grasp_img, arm_img, rh_mask])
            label_batch.append(label)
            if i_step%BATCH_SIZE==0:
                grasp_img_batch = np.array([grasp_img for grasp_img, arm_img, rh_mask in data_batch])
                arm_img_batch = np.array([arm_img for grasp_img, arm_img, rh_mask in data_batch])
                rh_mask_batch = np.array([rh_mask for grasp_img, arm_img, rh_mask in data_batch])
                label_batch = np.array(label_batch, dtype=np.int)
                train_step([grasp_img_batch, arm_img_batch, rh_mask_batch], label_batch)
                data_batch, label_batch = [], []
            if i_step%LOG_STEP==0:
                print("train step - {}/{}        ".format(i_step, N_train), end = '\r')
        with train_summary_writer.as_default():
            tf.summary.scalar('loss', train_loss.result(), step=epoch)
            tf.summary.scalar('accuracy', train_accuracy.result(), step=epoch)

        i_step = 0
        data_batch, label_batch = [], []
        for data_pair in data_pairs_test:
            i_step += 1
            grasp_img, arm_img, rh_mask, label = load_data(data_pair)
            data_batch.append([grasp_img, arm_img, rh_mask])
            label_batch.append(label)
            if i_step%BATCH_SIZE==0:
                grasp_img_batch = np.array([grasp_img for grasp_img, arm_img, rh_mask in data_batch])
                arm_img_batch = np.array([arm_img for grasp_img, arm_img, rh_mask in data_batch])
                rh_mask_batch = np.array([rh_mask for grasp_img, arm_img, rh_mask in data_batch])
                label_batch = np.array(label_batch, dtype=np.int)
                test_step([grasp_img_batch, arm_img_batch, rh_mask_batch], label_batch)
                data_batch, label_batch = [], []
            if i_step%LOG_STEP==0:
                print("test step - {}/{}        ".format(i_step, N_test), end = '\r')
        with test_summary_writer.as_default():
            tf.summary.scalar('loss', test_loss.result(), step=epoch)
            tf.summary.scalar('accuracy', test_accuracy.result(), step=epoch)

        model.save(model_log_dir.format(epoch + 1))

        print("")
        print("=================================================================")
        print(
            f'Epoch {epoch + 1}, '
            f'Loss: {train_loss.result()}, '
            f'Accuracy: {train_accuracy.result() * 100}, '
            f'Test Loss: {test_loss.result()}, '
            f'Test Accuracy: {test_accuracy.result() * 100}'
        )
        print("=================================================================")
        print("")
        res_list.append((float(test_accuracy.result())*100, float(test_loss.result())))
    return res_list

## train filtered data

In [8]:
ROBOT_MODEL_ROOT_DIV = "{}-filtered".format(ROBOT_MODEL_ROOT)
data_pairs_train_div, data_pairs_test_div = \
        data_pairs_train_filtered, data_pairs_test_filtered

res_list = train_script(data_pairs_train_div, data_pairs_test_div, ROBOT_MODEL_ROOT_DIV)
print("{} acc/loss list:".format(ROBOT_MODEL_ROOT_DIV))
for i_r, res in enumerate(res_list):
    print("{}: {} %, {} ".format(i_r+1, np.round(res[0],2), np.round(res[1], 3)))

Log path: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_1/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_1/assets



Epoch 1, Loss: 0.4174804091453552, Accuracy: 82.08000183105469, Test Loss: 0.5176162719726562, Test Accuracy: 75.5909423828125



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_2/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_2/assets



Epoch 2, Loss: 0.3408515453338623, Accuracy: 85.98999786376953, Test Loss: 0.7452200651168823, Test Accuracy: 71.03365325927734



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_3/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_3/assets



Epoch 3, Loss: 0.31742000579833984, Accuracy: 87.27999877929688, Test Loss: 0.3056412637233734, Test Accuracy: 87.61017608642578



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_4/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_4/assets



Epoch 4, Loss: 0.30118170380592346, Accuracy: 88.20000457763672, Test Loss: 0.447070449590683, Test Accuracy: 85.57691955566406



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_5/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_5/assets



Epoch 5, Loss: 0.28584185242652893, Accuracy: 88.42499542236328, Test Loss: 0.3372437655925751, Test Accuracy: 85.19631958007812



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_6/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_6/assets



Epoch 6, Loss: 0.277305543422699, Accuracy: 89.125, Test Loss: 0.29186761379241943, Test Accuracy: 87.890625



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_7/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_7/assets



Epoch 7, Loss: 0.2615160048007965, Accuracy: 89.40499877929688, Test Loss: 0.41178083419799805, Test Accuracy: 86.22796630859375



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_8/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_8/assets



Epoch 8, Loss: 0.26093992590904236, Accuracy: 89.95500183105469, Test Loss: 0.3198296129703522, Test Accuracy: 88.37139892578125



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_9/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_9/assets



Epoch 9, Loss: 0.24504604935646057, Accuracy: 90.72999572753906, Test Loss: 0.3289223313331604, Test Accuracy: 88.21113586425781



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_10/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_10/assets



Epoch 10, Loss: 0.22155888378620148, Accuracy: 90.97000122070312, Test Loss: 0.3273649513721466, Test Accuracy: 87.19952392578125



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_11/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_11/assets



Epoch 11, Loss: 0.2141284942626953, Accuracy: 92.22000122070312, Test Loss: 0.33303624391555786, Test Accuracy: 87.54006958007812



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_12/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_12/assets



Epoch 12, Loss: 0.18268676102161407, Accuracy: 93.11500549316406, Test Loss: 0.5442683100700378, Test Accuracy: 85.85737609863281



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_13/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_13/assets



Epoch 13, Loss: 0.17415669560432434, Accuracy: 93.44499969482422, Test Loss: 0.408098429441452, Test Accuracy: 87.93069458007812



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_14/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_14/assets



Epoch 14, Loss: 0.14899100363254547, Accuracy: 94.55500030517578, Test Loss: 0.3629663586616516, Test Accuracy: 86.61859130859375



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_15/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-filtered/20210923-144309/model_15/assets



Epoch 15, Loss: 0.1428585648536682, Accuracy: 94.7300033569336, Test Loss: 0.36193662881851196, Test Accuracy: 85.84735107421875

/home/rnb/Projects/rnb-planning/model/latticized/panda-filtered acc/loss list:
1: 75.59 %, 0.518 
2: 71.03 %, 0.745 
3: 87.61 %, 0.306 
4: 85.58 %, 0.447 
5: 85.2 %, 0.337 
6: 87.89 %, 0.292 
7: 86.23 %, 0.412 
8: 88.37 %, 0.32 
9: 88.21 %, 0.329 
10: 87.2 %, 0.327 
11: 87.54 %, 0.333 
12: 85.86 %, 0.544 
13: 87.93 %, 0.408 
14: 86.62 %, 0.363 
15: 85.85 %, 0.362 


## failmore data

In [9]:
ROBOT_MODEL_ROOT_DIV = "{}-failmore".format(ROBOT_MODEL_ROOT)
data_pairs_train_div, data_pairs_test_div = \
        data_pairs_train_failmore, data_pairs_test_failmore

res_list = train_script(data_pairs_train_div, data_pairs_test_div, ROBOT_MODEL_ROOT_DIV)
print("{} acc/loss list:".format(ROBOT_MODEL_ROOT_DIV))
for i_r, res in enumerate(res_list):
    print("{}: {} %, {} ".format(i_r+1, np.round(res[0],2), np.round(res[1], 3)))

Log path: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_1/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_1/assets



Epoch 1, Loss: 0.24209104478359222, Accuracy: 93.9800033569336, Test Loss: 0.08203324675559998, Test Accuracy: 96.13381958007812



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_2/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_2/assets



Epoch 2, Loss: 0.0714544728398323, Accuracy: 97.19499969482422, Test Loss: 0.07664866745471954, Test Accuracy: 96.85496520996094



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_3/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_3/assets



Epoch 3, Loss: 0.06051875650882721, Accuracy: 97.62999725341797, Test Loss: 0.08596449345350266, Test Accuracy: 96.484375



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_4/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_4/assets



Epoch 4, Loss: 0.057302460074424744, Accuracy: 97.84500122070312, Test Loss: 0.1053919792175293, Test Accuracy: 94.87179565429688



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_5/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_5/assets



Epoch 5, Loss: 0.05527128279209137, Accuracy: 97.68999481201172, Test Loss: 0.0677516981959343, Test Accuracy: 96.93509674072266



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_6/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_6/assets



Epoch 6, Loss: 0.048642709851264954, Accuracy: 98.09500122070312, Test Loss: 0.07656817883253098, Test Accuracy: 96.91506958007812



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_7/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_7/assets



Epoch 7, Loss: 0.04965522512793541, Accuracy: 98.11000061035156, Test Loss: 0.08616393804550171, Test Accuracy: 96.2940673828125



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_8/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_8/assets



Epoch 8, Loss: 0.04370955005288124, Accuracy: 98.32500457763672, Test Loss: 0.07638297230005264, Test Accuracy: 96.93509674072266



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_9/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_9/assets



Epoch 9, Loss: 0.042307768017053604, Accuracy: 98.33499908447266, Test Loss: 0.07246807217597961, Test Accuracy: 97.05528259277344



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_10/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_10/assets



Epoch 10, Loss: 0.03637075051665306, Accuracy: 98.68500518798828, Test Loss: 0.09645368903875351, Test Accuracy: 96.6846923828125



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_11/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_11/assets



Epoch 11, Loss: 0.037487562745809555, Accuracy: 98.46500396728516, Test Loss: 0.10753750056028366, Test Accuracy: 95.64302825927734



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_12/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_12/assets



Epoch 12, Loss: 0.034505486488342285, Accuracy: 98.58499908447266, Test Loss: 0.08839751034975052, Test Accuracy: 96.84495544433594



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_13/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_13/assets



Epoch 13, Loss: 0.033748213201761246, Accuracy: 98.90499877929688, Test Loss: 0.11252864450216293, Test Accuracy: 96.875



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_14/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_14/assets



Epoch 14, Loss: 0.03798489272594452, Accuracy: 98.63500213623047, Test Loss: 0.08285766839981079, Test Accuracy: 97.08534240722656



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_15/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-failmore/20210923-150108/model_15/assets



Epoch 15, Loss: 0.02693922072649002, Accuracy: 99.04500579833984, Test Loss: 0.09839364141225815, Test Accuracy: 96.61457824707031

/home/rnb/Projects/rnb-planning/model/latticized/panda-failmore acc/loss list:
1: 96.13 %, 0.082 
2: 96.85 %, 0.077 
3: 96.48 %, 0.086 
4: 94.87 %, 0.105 
5: 96.94 %, 0.068 
6: 96.92 %, 0.077 
7: 96.29 %, 0.086 
8: 96.94 %, 0.076 
9: 97.06 %, 0.072 
10: 96.68 %, 0.096 
11: 95.64 %, 0.108 
12: 96.84 %, 0.088 
13: 96.88 %, 0.113 
14: 97.09 %, 0.083 
15: 96.61 %, 0.098 


## balanced data

In [10]:
ROBOT_MODEL_ROOT_DIV = "{}-balanced".format(ROBOT_MODEL_ROOT)
data_pairs_train_div, data_pairs_test_div = \
        data_pairs_train_mixed, data_pairs_test_mixed

res_list = train_script(data_pairs_train_div, data_pairs_test_div, ROBOT_MODEL_ROOT_DIV)
print("{} acc/loss list:".format(ROBOT_MODEL_ROOT_DIV))
for i_r, res in enumerate(res_list):
    print("{}: {} %, {} ".format(i_r+1, np.round(res[0],2), np.round(res[1], 3)))

Log path: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_1/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_1/assets



Epoch 1, Loss: 0.16795837879180908, Accuracy: 94.5999984741211, Test Loss: 0.1439146250486374, Test Accuracy: 95.19230651855469



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_2/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_2/assets



Epoch 2, Loss: 0.09621457755565643, Accuracy: 96.73500061035156, Test Loss: 0.11894498765468597, Test Accuracy: 96.103759765625



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_3/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_3/assets



Epoch 3, Loss: 0.08410865813493729, Accuracy: 96.95500183105469, Test Loss: 0.1484803408384323, Test Accuracy: 95.60296630859375



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_4/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_4/assets



Epoch 4, Loss: 0.07246683537960052, Accuracy: 97.31999969482422, Test Loss: 0.13625557720661163, Test Accuracy: 95.5028076171875



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_5/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_5/assets



Epoch 5, Loss: 0.06380670517683029, Accuracy: 97.95500183105469, Test Loss: 0.1327686607837677, Test Accuracy: 96.14382934570312



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_6/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_6/assets



Epoch 6, Loss: 0.05716915428638458, Accuracy: 98.03500366210938, Test Loss: 0.14565008878707886, Test Accuracy: 96.083740234375



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_7/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_7/assets



Epoch 7, Loss: 0.04987991973757744, Accuracy: 98.4000015258789, Test Loss: 0.32462841272354126, Test Accuracy: 88.43148803710938



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_8/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_8/assets



Epoch 8, Loss: 0.0532381646335125, Accuracy: 98.30500030517578, Test Loss: 0.1337578296661377, Test Accuracy: 95.78324890136719



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_9/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_9/assets



Epoch 9, Loss: 0.027271978557109833, Accuracy: 99.1050033569336, Test Loss: 0.143043652176857, Test Accuracy: 96.103759765625



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_10/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_10/assets



Epoch 10, Loss: 0.026352355256676674, Accuracy: 99.08999633789062, Test Loss: 0.18549920618534088, Test Accuracy: 95.22235107421875



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_11/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_11/assets



Epoch 11, Loss: 0.03758059814572334, Accuracy: 98.70999908447266, Test Loss: 0.1928151398897171, Test Accuracy: 94.43109130859375



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_12/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_12/assets



Epoch 12, Loss: 0.043983619660139084, Accuracy: 98.5999984741211, Test Loss: 0.1599022001028061, Test Accuracy: 95.703125



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_13/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_13/assets



Epoch 13, Loss: 0.0248726774007082, Accuracy: 99.1050033569336, Test Loss: 0.17157118022441864, Test Accuracy: 95.94351196289062



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_14/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_14/assets



Epoch 14, Loss: 0.019831575453281403, Accuracy: 99.2750015258789, Test Loss: 0.18808874487876892, Test Accuracy: 95.56289672851562



INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_15/assets


INFO:tensorflow:Assets written to: /home/rnb/Projects/rnb-planning/model/latticized/panda-balanced/20210923-151926/model_15/assets



Epoch 15, Loss: 0.019988996908068657, Accuracy: 99.44999694824219, Test Loss: 0.1989697813987732, Test Accuracy: 96.06370544433594

/home/rnb/Projects/rnb-planning/model/latticized/panda-balanced acc/loss list:
1: 95.19 %, 0.144 
2: 96.1 %, 0.119 
3: 95.6 %, 0.148 
4: 95.5 %, 0.136 
5: 96.14 %, 0.133 
6: 96.08 %, 0.146 
7: 88.43 %, 0.325 
8: 95.78 %, 0.134 
9: 96.1 %, 0.143 
10: 95.22 %, 0.185 
11: 94.43 %, 0.193 
12: 95.7 %, 0.16 
13: 95.94 %, 0.172 
14: 95.56 %, 0.188 
15: 96.06 %, 0.199 


# Train results - Panda

#### filtered: 20210923-144309
```
1: 75.59 %, 0.518 
2: 71.03 %, 0.745 
3: 87.61 %, 0.306 
4: 85.58 %, 0.447 
5: 85.2 %, 0.337 
6: 87.89 %, 0.292 
7: 86.23 %, 0.412 
8: 88.37 %, 0.32 
9: 88.21 %, 0.329 
10: 87.2 %, 0.327 
11: 87.54 %, 0.333 
12: 85.86 %, 0.544 
13: 87.93 %, 0.408 
14: 86.62 %, 0.363 
15: 85.85 %, 0.362 
```

#### failmore: 20210923-150108
```
1: 96.13 %, 0.082 
2: 96.85 %, 0.077 
3: 96.48 %, 0.086 
4: 94.87 %, 0.105 
5: 96.94 %, 0.068 
6: 96.92 %, 0.077 
7: 96.29 %, 0.086 
8: 96.94 %, 0.076 
9: 97.06 %, 0.072 
10: 96.68 %, 0.096 
11: 95.64 %, 0.108 
12: 96.84 %, 0.088 
13: 96.88 %, 0.113 
14: 97.09 %, 0.083 
15: 96.61 %, 0.098 
```

#### balanced: 20210610-113433
```
1: 95.19 %, 0.144 
2: 96.1 %, 0.119 
3: 95.6 %, 0.148 
4: 95.5 %, 0.136 
5: 96.14 %, 0.133 
6: 96.08 %, 0.146 
7: 88.43 %, 0.325 
8: 95.78 %, 0.134 
9: 96.1 %, 0.143 
10: 95.22 %, 0.185 
11: 94.43 %, 0.193 
12: 95.7 %, 0.16 
13: 95.94 %, 0.172 
14: 95.56 %, 0.188 
15: 96.06 %, 0.199 
```

## New train 2021-06
* **[IMPORTANT] 이전에 Tej 사용 틀린 점 보완 했음** 

### Panda
* 20210606-080139 : first balanced train trial, 96.6% 달성 / 이전 데이터와 collision boundary shape 차이일 수 있음
* 20210606-082302 : 이전 데이터와 전부 섞어서 collision boundary shape 차이 영향 확인 - 87% 가량, 영향 확실, 재수집 필요
* 20210606-191455 : 새로 모았어도 분리 학습 하니 시작부터 99%, 문제 있음
* 20210606-194148 : 새로 모은거 분리 학습, 테스트만 순수 기존 데이터로 하니 59%, 문제 확실
* 20210606-201525 : 새 데이터만 학습, 96%? 하지만 fail bias 돼 있음.
* 20210606-204309 : 새 데이터만 학습, test set은 bias 제거함 ep8 89% 수준

### Indy7
* 20210607-224401 : 새 데이터만 학습, test set은 bias 제거함 ep12 79.5% 수준
* 20210610-052216 : 조인트 리밋 해제한 데이터만 학습, test set은 bias 제거, ep12 92% 수준

## Load & test

In [8]:
def get_tf_functions(model_root, model_folder, epoch_folder):
    # last_model = '20210606-204309' # panda
    # last_model = '20210610-052216' # indy
    # last_model = sorted(os.listdir(ROBOT_MODEL_ROOT))[-1]
    logpath = os.path.join(model_root, model_folder)

    # epochlist = sorted([item for item in os.listdir(logpath) if item.startswith("model")], key=lambda x: int(x[6:]))
    # print(epochlist)
    # last_save = epochlist[-1]
    model_log_dir = os.path.join(logpath, epoch_folder)

    import tensorflow as tf
    model = tf.keras.models.load_model(model_log_dir)

    @tf.function
    def inference(images):
        # training=False is only needed if there are layers with different
        # behavior during training versus inference (e.g. Dropout).
        predictions = model(images, training=False)
        return predictions

    loss_object = tf.keras.losses.BinaryCrossentropy()

    @tf.function
    def calc_loss(labels, predictions):
        # training=False is only needed if there are layers with different
        # behavior during training versus inference (e.g. Dropout).
        return loss_object(labels, predictions)
    return model, inference, calc_loss

def test_script(data_pairs_test, model_root, model_folder, epoch_folder):
    BATCH_SIZE = 1
    LOG_STEP = 100
    N_test = len(data_pairs_test)
    gtimer = GlobalTimer.instance()
    gtimer.reset()
    
    model, inference, calc_loss= get_tf_functions(model_root, model_folder, epoch_folder)

    grasp_img, arm_img, rh_mask, label = load_data(data_pairs_test[0])
    res = inference([np.array([grasp_img]), np.array([arm_img]), np.array([rh_mask])])


    i_step = 0
    res_list = []
    label_list = []
    loss_list= []
    grasp_img_batch = []
    arm_img_batch = []
    rh_mask_batch = []
    label_batch = [] 
    for data_pair in data_pairs_test:
        i_step += 1
        grasp_img, arm_img, rh_mask, label = load_data(data_pair)
        grasp_img_batch.append(grasp_img)
        arm_img_batch.append(arm_img)
        rh_mask_batch.append(rh_mask)
        label_batch.append(label)
        if len(grasp_img_batch)==BATCH_SIZE:
            grasp_img_batch, arm_img_batch, rh_mask_batch = np.array(grasp_img_batch), np.array(arm_img_batch), np.array(rh_mask_batch)
            with gtimer.block("inference"):
                res = inference([grasp_img_batch, arm_img_batch, rh_mask_batch])
            loss = calc_loss(label_batch, res)
            res_list = res_list + list(res.numpy()>0.5)
            label_list = label_list + label_batch
            loss_list.append(loss.numpy())
            grasp_img_batch = []
            arm_img_batch = []
            rh_mask_batch = []
            label_batch = [] 
        if i_step%LOG_STEP==0:
            print("test step - {}/{}        ".format(i_step, N_test), end = '\r')

    res_list = np.array(res_list)[:,1]
    label_list = np.array(label_list)[:,1]
    loss_list = np.array(loss_list)[:]

    acc = np.mean(np.equal(res_list, label_list)) * 100
    mean_loss = np.mean(loss_list)

    print("")
    print("=================================================================")
    print(
        f'Test Loss: {mean_loss} \n'
        f'Test Accuracy: {acc} \n'
        f'TP / FN / ACC: {np.sum(np.logical_and(res_list, label_list))}, ' 
        f'{np.sum(np.logical_and(np.logical_not(res_list), label_list))}, ' 
        f'{round(np.mean(res_list[np.where(label_list)])*100,2)}\n'
        f'FP / TN / ACC: {np.sum(np.logical_and(res_list, np.logical_not(label_list)))}, '
        f'{np.sum(np.logical_and(np.logical_not(res_list), np.logical_not(label_list)))}, '
        f'{round(np.mean(np.logical_not(res_list[np.where(np.logical_not(label_list))]))*100,2)}\n'
        f'PACC / NACC / TACC: {round(np.mean(label_list[np.where(res_list)])*100,2)}, '
        f'{round(np.mean(np.logical_not(label_list[np.where(np.logical_not(res_list))]))*100,2)}, '
        f'{round(np.mean(res_list==label_list)*100,2)}\n'
    )
    print("=================================================================")
    print("")
    print(gtimer)

    TP_FN_FP_TN = (np.sum(np.logical_and(res_list, label_list)), np.sum(np.logical_and(np.logical_not(res_list), label_list)), 
                   np.sum(np.logical_and(res_list, np.logical_not(label_list))), np.sum(np.logical_and(np.logical_not(res_list), np.logical_not(label_list))))
        
    return label_list, res_list

# Test all

In [9]:
print("========= Filtered =========")
model_foler, epoch_folder = '20210610-104017', 'model_4'
filtered_res = []
filtered_res.append(test_script(data_pairs_test_filtered, ROBOT_MODEL_ROOT+"-filtered", model_foler, epoch_folder))
filtered_res.append(test_script(data_pairs_test_failmore, ROBOT_MODEL_ROOT+"-filtered", model_foler, epoch_folder))
filtered_res.append(test_script(data_pairs_test_mixed, ROBOT_MODEL_ROOT+"-filtered", model_foler, epoch_folder))

print("========= Failmore =========")
model_foler, epoch_folder = '20210610-110737', 'model_15'
failmore_res = []
failmore_res.append(test_script(data_pairs_test_filtered, ROBOT_MODEL_ROOT+"-failmore", model_foler, epoch_folder))
failmore_res.append(test_script(data_pairs_test_failmore, ROBOT_MODEL_ROOT+"-failmore", model_foler, epoch_folder))
failmore_res.append(test_script(data_pairs_test_mixed, ROBOT_MODEL_ROOT+"-failmore", model_foler, epoch_folder))

print("========= Balanced =========")
model_foler, epoch_folder = '20210610-113433', 'model_15'
balanced_res = []
balanced_res.append(test_script(data_pairs_test_filtered, ROBOT_MODEL_ROOT+"-balanced", model_foler, epoch_folder))
balanced_res.append(test_script(data_pairs_test_failmore, ROBOT_MODEL_ROOT+"-balanced", model_foler, epoch_folder))
balanced_res.append(test_script(data_pairs_test_mixed, ROBOT_MODEL_ROOT+"-balanced", model_foler, epoch_folder))

print("========= Filtered =========")
print("filtered:")
print_confusion_mat(*filtered_res[0])
print("failmore:")
print_confusion_mat(*filtered_res[1])
print("balanced:")
print_confusion_mat(*filtered_res[2])
print("========================\n")

print("========= Failmore =========")
print("filtered:")
print_confusion_mat(*failmore_res[0])
print("failmore:")
print_confusion_mat(*failmore_res[1])
print("balanced:")
print_confusion_mat(*failmore_res[2])
print("========================\n")

print("========= Balanced =========")
print("filtered:")
print_confusion_mat(*balanced_res[0])
print("failmore:")
print_confusion_mat(*balanced_res[1])
print("balanced:")
print_confusion_mat(*balanced_res[2])
print("========================\n")

========= Filtered =========


test step - 5000/5000        
Test Loss: 0.3913080394268036 
Test Accuracy: 85.2 
TP / FN / ACC: 2673, 252, 91.38
FP / TN / ACC: 488, 1587, 76.48
PACC / NACC / TACC: 84.56, 86.3, 85.2


inference: 	127565.0 ms/5000 = 25.513 ms (10.381/41.004)



test step - 5000/5000        
Test Loss: 0.8329949378967285 
Test Accuracy: 68.36 
TP / FN / ACC: 172, 9, 95.03
FP / TN / ACC: 1573, 3246, 67.36
PACC / NACC / TACC: 9.86, 99.72, 68.36


inference: 	127350.0 ms/5000 = 25.47 ms (11.811/30.964)



test step - 5000/5000        
Test Loss: 0.5485051870346069 
Test Accuracy: 79.66 
TP / FN / ACC: 2273, 227, 90.92
FP / TN / ACC: 790, 1710, 68.4
PACC / NACC / TACC: 74.21, 88.28, 79.66


inference: 	127729.0 ms/5000 = 25.546 ms (10.221/31.68)

========= Failmore =========


test step - 5000/5000        
Test Loss: 1.454448938369751 
Test Accuracy: 66.46 
TP / FN / ACC: 1338, 1587, 45.74
FP / TN / ACC: 90, 1985, 95.66
PACC / NACC / TACC: 93.7, 55.57, 66.46


inference: 	128127.0 ms/5000 = 25.625 ms (11.684/31.42)



test step - 5000/5000        
Test Loss: 0.07828611880540848 
Test Accuracy: 97.58 
TP / FN / ACC: 104, 77, 57.46
FP / TN / ACC: 44, 4775, 99.09
PACC / NACC / TACC: 70.27, 98.41, 97.58


inference: 	128798.0 ms/5000 = 25.76 ms (12.115/31.827)



test step - 5000/5000        
Test Loss: 1.1915452480316162 
Test Accuracy: 72.64 
TP / FN / ACC: 1158, 1342, 46.32
FP / TN / ACC: 26, 2474, 98.96
PACC / NACC / TACC: 97.8, 64.83, 72.64


inference: 	128689.0 ms/5000 = 25.738 ms (12.217/33.968)

========= Balanced =========


test step - 5000/5000        
Test Loss: 4.108362197875977 
Test Accuracy: 58.74 
TP / FN / ACC: 2925, 0, 100.0
FP / TN / ACC: 2063, 12, 0.58
PACC / NACC / TACC: 58.64, 100.0, 58.74


inference: 	128773.0 ms/5000 = 25.755 ms (12.782/31.845)



test step - 5000/5000        
Test Loss: 0.30704060196876526 
Test Accuracy: 96.3 
TP / FN / ACC: 1, 180, 0.55
FP / TN / ACC: 5, 4814, 99.9
PACC / NACC / TACC: 16.67, 96.4, 96.3


inference: 	128515.0 ms/5000 = 25.703 ms (12.191/31.163)



test step - 5000/5000        
Test Loss: 0.001708336640149355 
Test Accuracy: 99.92 
TP / FN / ACC: 2500, 0, 100.0
FP / TN / ACC: 4, 2496, 99.84
PACC / NACC / TACC: 99.84, 100.0, 99.92


inference: 	128326.0 ms/5000 = 25.665 ms (11.319/31.504)

========= Filtered =========
filtered:
	 PP 	 	 PN 	 	 5000
GP 	 2673 	 	 252 	 	 91.38%
GN 	 488 	 	 1587 	 76.48%
AL 	 84.56% 	 86.30% 	 85.20%
failmore:
	 PP 	 	 PN 	 	 5000
GP 	 172 	 	 9 	 	 95.03%
GN 	 1573 	 	 3246 	 67.36%
AL 	 9.86% 	 99.72% 	 68.36%
balanced:
	 PP 	 	 PN 	 	 5000
GP 	 2273 	 	 227 	 	 90.92%
GN 	 790 	 	 1710 	 68.40%
AL 	 74.21% 	 88.28% 	 79.66%

========= Failmore =========
filtered:
	 PP 	 	 PN 	 	 5000
GP 	 1338 	 	 1587 	 	 45.74%
GN 	 90 	 	 1985 	 95.66%
AL 	 93.70% 	 55.57% 	 66.46%
failmore:
	 PP 	 	 PN 	 	 5000
GP 	 104 	 	 77 	 	 57.46%
GN 	 44 	 	 4775 	 99.09%
AL 	 70.27% 	 98.41% 	 97.58%
balanced:
	 PP 	 	 PN 	 	 5000
GP 	 1158 	 	 1342 	 	 46.32%
GN 	 26 	 	 2474 	 98.96%
AL 	 97.80% 	 64.83% 	 72.64%


### Test result - panda
```
inference: 	128326.0 ms/5000 = 25.665 ms (11.319/31.504)

========= Filtered =========
filtered:
	 PP 	 	 PN 	 	 5000
GP 	 2673 	 	 252 	 	 91.38%
GN 	 488 	 	 1587 	 76.48%
AL 	 84.56% 	 86.30% 	 85.20%
failmore:
	 PP 	 	 PN 	 	 5000
GP 	 172 	 	 9 	 	 95.03%
GN 	 1573 	 	 3246 	 67.36%
AL 	 9.86% 	 99.72% 	 68.36%
balanced:
	 PP 	 	 PN 	 	 5000
GP 	 2273 	 	 227 	 	 90.92%
GN 	 790 	 	 1710 	 68.40%
AL 	 74.21% 	 88.28% 	 79.66%
========================

========= Failmore =========
filtered:
	 PP 	 	 PN 	 	 5000
GP 	 1338 	 	 1587 	 	 45.74%
GN 	 90 	 	 1985 	 95.66%
AL 	 93.70% 	 55.57% 	 66.46%
failmore:
	 PP 	 	 PN 	 	 5000
GP 	 104 	 	 77 	 	 57.46%
GN 	 44 	 	 4775 	 99.09%
AL 	 70.27% 	 98.41% 	 97.58%
balanced:
	 PP 	 	 PN 	 	 5000
GP 	 1158 	 	 1342 	 	 46.32%
GN 	 26 	 	 2474 	 98.96%
AL 	 97.80% 	 64.83% 	 72.64%
========================

========= Balanced =========
filtered:
	 PP 	 	 PN 	 	 5000
GP 	 2925 	 	 0 	 	 100.00%
GN 	 2063 	 	 12 	 0.58%
AL 	 58.64% 	 100.00% 	 58.74%
failmore:
	 PP 	 	 PN 	 	 5000
GP 	 1 	 	 180 	 	 0.55%
GN 	 5 	 	 4814 	 99.90%
AL 	 16.67% 	 96.40% 	 96.30%
balanced:
	 PP 	 	 PN 	 	 5000
GP 	 2500 	 	 0 	 	 100.00%
GN 	 4 	 	 2496 	 99.84%
AL 	 99.84% 	 100.00% 	 99.92%
========================
```

## Test convergence

In [ ]:
last_model = sorted(os.listdir(ROBOT_MODEL_ROOT))[-1]
logpath = os.path.join(ROBOT_MODEL_ROOT, last_model)

model_epoch_list = []
acc_epoch_list = []
loss_epoch_list = []
# last_save = sorted([item for item in os.listdir(logpath) if item.startswith("model")])[-1]
# last_save = 'model_1'
for last_save in sorted([item for item in os.listdir(logpath) if item.startswith("model")], key=lambda x: int(x[6:])):
    model_log_dir = os.path.join(logpath, last_save)

    import tensorflow as tf
    model = tf.keras.models.load_model(model_log_dir)

    @tf.function
    def inference(images):
        # training=False is only needed if there are layers with different
        # behavior during training versus inference (e.g. Dropout).
        predictions = model(images, training=False)
        return predictions

    loss_object = tf.keras.losses.BinaryCrossentropy()

    @tf.function
    def calc_loss(labels, predictions):
        # training=False is only needed if there are layers with different
        # behavior during training versus inference (e.g. Dropout).
        return loss_object(labels, predictions)
    
    BATCH_SIZE = 50
    LOG_STEP = 100
    N_test = len(data_pairs_test)
    gtimer = GlobalTimer.instance()
    gtimer.reset()

    i_step = 0
    res_list = []
    label_list = []
    loss_list= []
    grasp_img_batch = []
    arm_img_batch = []
    rh_mask_batch = []
    label_batch = [] 
    for data_pair in data_pairs_test:
        i_step += 1
        grasp_img, arm_img, rh_mask, label = load_data(data_pair)
        grasp_img_batch.append(grasp_img)
        arm_img_batch.append(arm_img)
        rh_mask_batch.append(rh_mask)
        label_batch.append(label)
        if len(grasp_img_batch)==BATCH_SIZE:
            grasp_img_batch, arm_img_batch, rh_mask_batch = np.array(grasp_img_batch), np.array(arm_img_batch), np.array(rh_mask_batch)
            with gtimer.block("inference"):
                res = inference([grasp_img_batch, arm_img_batch, rh_mask_batch])
            loss = calc_loss(label_batch, res)
            res_list = res_list + list(res.numpy()>0.5)
            label_list = label_list + label_batch
            loss_list.append(loss.numpy())
            grasp_img_batch = []
            arm_img_batch = []
            rh_mask_batch = []
            label_batch = [] 
        if i_step%LOG_STEP==0:
            print("test step - {}/{}        ".format(i_step, N_test), end = '\r')

    res_list = np.array(res_list)[:5000,1]
    label_list = np.array(label_list)[:5000,1]
    loss_list = np.array(loss_list)[:5000]
    
    acc = np.mean(np.equal(res_list, label_list)) * 100
    mean_loss = np.mean(loss_list)

    print("")
    print("=================================================================")
    print(
        f'Test Loss: {mean_loss} \n'
        f'Test Accuracy: {acc} \n'
        f'TP / FN / ACC: {np.sum(np.logical_and(res_list, label_list))}, ' 
        f'{np.sum(np.logical_and(np.logical_not(res_list), label_list))}, ' 
        f'{round(np.mean(res_list[np.where(label_list)])*100,2)}\n'
        f'FP / TN / ACC: {np.sum(np.logical_and(res_list, np.logical_not(label_list)))}, '
        f'{np.sum(np.logical_and(np.logical_not(res_list), np.logical_not(label_list)))}, '
        f'{round(np.mean(np.logical_not(res_list[np.where(np.logical_not(label_list))]))*100,2)}\n'
        f'PACC / NACC / TACC: {round(np.mean(label_list[np.where(res_list)])*100,2)}, '
        f'{round(np.mean(np.logical_not(label_list[np.where(np.logical_not(res_list))]))*100,2)}, '
        f'{round(np.mean(res_list==label_list)*100,2)}\n'
    )
    print("=================================================================")
    print("")
    print(gtimer)
    model_epoch_list.append(last_save)
    acc_epoch_list.append(acc)
    loss_epoch_list.append(mean_loss)

In [ ]:
import matplotlib.pyplot as plt
print(model_epoch_list)
plt.figure(figsize=(10,3))
plt.subplot(1,2,1)
plt.plot(acc_epoch_list)
plt.subplot(1,2,2)
plt.plot(loss_epoch_list)

In [ ]:
save_json(ROBOT_TYPE_NAME+".json", {"epoch": np.array(model_epoch_list), "acc": np.array(acc_epoch_list), "loss": np.array(loss_epoch_list)})

## test on shared array

In [ ]:
import SharedArray as sa
BATCH_SIZE = 1
SERVER_PERIOD = 1e-2
# Create an array in shared memory.
robot_type_p = sa.attach("shm://robot_type")
grasp_img_p = sa.attach("shm://grasp_img")
arm_img_p = sa.attach("shm://arm_img")
rh_mask_p = sa.attach("shm://rh_mask")
result_p = sa.attach("shm://result")
query_in = sa.attach("shm://query_in")
response_out = sa.attach("shm://response_out")
query_quit = sa.attach("shm://query_quit")

def query_wait_response(grasp_img_batch, arm_img_batch, rh_mask_batch):
    grasp_img_p[:] = grasp_img_batch[:]
    arm_img_p[:] = arm_img_batch[:]
    rh_mask_p[:] = rh_mask_batch[:]
    query_in[0] = True
    while not response_out[0]:
        time.sleep(SERVER_PERIOD)
    response_out[0] = False
    return np.copy(result_p)

def quit_shared_server():
    query_quit[0] = True

In [ ]:
LOG_STEP = 100
N_test = len(data_pairs_test)

i_step = 0
data_batch, label_batch = [], []
result_list = []
label_list = []
for data_pair in data_pairs_test:
    i_step += 1
    grasp_img, arm_img, rh_mask, label = load_data(data_pair)
    data_batch.append([grasp_img, arm_img, rh_mask])
    label_batch.append(label)
    if i_step%BATCH_SIZE==0:
        grasp_img_batch = np.array([grasp_img for grasp_img, arm_img, rh_mask in data_batch])
        arm_img_batch = np.array([arm_img for grasp_img, arm_img, rh_mask in data_batch])
        rh_mask_batch = np.array([rh_mask for grasp_img, arm_img, rh_mask in data_batch])
        label_batch = np.array(label_batch, dtype=np.int)
        result = query_wait_response(grasp_img_batch, arm_img_batch, rh_mask_batch)
        result_list.append(result)
        label_list.append(label_batch)
        data_batch, label_batch = [], []
    if i_step%LOG_STEP==0:
        print("test step - {}/{}        ".format(i_step, N_test), end = '\r')

In [ ]:
result_all = np.array(result_list)
label_all = np.array(label_list)

In [ ]:
np.mean(np.round(result_all).astype(np.int) == label_all.astype(np.int), axis=0)

In [ ]:
quit_shared_server()

## save gridded data

In [ ]:
data_pair = data_pairs_test[0]
grasp_data = load_pickle(data_pair[0])
arm_data = load_pickle(data_pair[1])
grasp_tar_idx = grasp_data[b'tar']
grasp_tool_idx = grasp_data[b'tool']
arm_tar_idx = arm_data[b'tar']
Tee = grasp_data[b'T_end_effector']
Tej = grasp_data[b'T_end_joint']
Tref_base = grasp_data[b'Tref_base']
reach_lb = grasp_data[b'reach']
retrieve_lb = grasp_data[b'retrieve']
r, th, h = cart2cyl(*Tee[:3,3])
r_ej, th, h_ej = cart2cyl(*Tej[:3,3])
r_class = div_r(r_ej)
h_class = div_h(h_ej)
r_mask = np.zeros(RH_MASK_SIZE)
r_mask[r_class*RH_MASK_STEP:r_class*RH_MASK_STEP+RH_MASK_STEP] = 1
h_mask = np.zeros(RH_MASK_SIZE)
h_mask[h_class*RH_MASK_STEP:h_class*RH_MASK_STEP+RH_MASK_STEP] = 1
rh_mask = np.concatenate([r_mask, h_mask])
# r_ej_list.append(r_ej)
# h_ej_list.append(h_ej)
# reach_lb_list.append(reach_lb)
Tref = SE3(Rot_axis(3, th), Tee[:3,3])
grasp_tool_img = np.zeros(GRASP_SHAPE)
grasp_tar_img = np.zeros(GRASP_SHAPE)
grasp_tool_img[np.unravel_index(grasp_tool_idx, shape=GRASP_SHAPE)] = 1
grasp_tar_img[np.unravel_index(grasp_tar_idx, shape=GRASP_SHAPE)] = 1
arm_img = np.zeros(ARM_SHAPE+(1,))
arm_img[np.unravel_index(arm_tar_idx, shape=ARM_SHAPE)] = 1
grasp_img = np.stack([grasp_tool_img, grasp_tar_img], axis=-1)

In [ ]:
# save_json("tar_arm.json", np.array(np.unravel_index(arm_tar_idx, shape=ARM_SHAPE)).transpose())

In [ ]:
# save_json("tool_effector.json", np.array(np.unravel_index(grasp_tool_idx, shape=GRASP_SHAPE)).transpose())
# save_json("tar_effector.json", np.array(np.unravel_index(grasp_tar_idx, shape=GRASP_SHAPE)).transpose())

## visualize r, h class distribution

In [ ]:
import matplotlib.pyplot as plt
r_ej_list = []
h_ej_list = []
for data_pair in data_pairs_train:
    grasp_data = load_pickle(data_pair[0])
    arm_data = load_pickle(data_pair[1])
    grasp_obj_idx = grasp_data[b'obj']
    grasp_tar_idx = grasp_data[b'tar']
    grasp_tool_idx = grasp_data[b'tool']
    arm_tar_idx = arm_data[b'tar']
    Tee = grasp_data[b'T_end_effector']
    Tej = grasp_data[b'T_end_joint']
    Tref_base = grasp_data[b'Tref_base']
    reach_lb = grasp_data[b'reach']
    retrieve_lb = grasp_data[b'retrieve']
    r, th, h = cart2cyl(*Tee[:3,3])
    r_ej, th, h_ej = cart2cyl(*Tej[:3,3])
    r_ej_list.append(r_ej)
    h_ej_list.append(h_ej)

In [ ]:
def div_r(r):
    return floor(sigmoid((r)/0.1-7)*8)

def div_h(h):
    return floor(sigmoid((h+0.6)/0.2-5)*8)

In [ ]:
plt.subplot(1,2,1)
plt.plot(sorted([div_r(r_ej_) for r_ej_ in r_ej_list]), '.')
plt.subplot(1,2,2)
plt.plot(sorted([div_h(h_ej_) for h_ej_ in h_ej_list]), '.')

In [ ]:
plt.subplot(1,2,1)
plt.plot(sorted([div_r(r_ej_) for r_ej_ in r_ej_list]), '.')
plt.subplot(1,2,2)
plt.plot(sorted([div_h(h_ej_) for h_ej_ in h_ej_list]), '.')

In [ ]:

plt.plot(sorted(np.floor(np.divide(r_ej_list,0.1))), '.')
plt.plot(sorted(np.floor(np.divide(h_ej_list,0.1))), '.')

In [ ]:

plt.plot(sorted(r_ej_list), '.')
plt.plot(sorted(h_ej_list), '.')

In [ ]:
np.max(r_ej_list)

In [ ]:
1.0 - 0.2

In [ ]:
np.min((np.array(r_ej_list)-0.2))

In [ ]:
np.min(np.floor((np.array(r_ej_list)-0.2)/0.05).astype(np.int))

In [ ]:
def gaussian(x, mu, sig):
    return np.exp(-np.power(x - mu, 2.) / (2 * np.power(sig, 2.)))

In [ ]:
x = div_r_gaussian(sorted(r_ej_list)[5000])

In [ ]:
y = div_h_gaussian(sorted(h_ej_list)[5000])

In [ ]:
plt.figure(figsize=(11,5))
plt.imshow(x[np.newaxis, :], cmap='gray')

In [ ]:
plt.figure(figsize=(16,5))
plt.imshow(y[np.newaxis, :], cmap='gray')

In [ ]:
def div_h_gaussian(h_val):
    return gaussian(h_val, np.arange(-0.5,1.1, 0.05),0.05)

In [ ]:

def div_h(h):
    return floor(sigmoid((h+0.6)/0.2-4.5)*8)

In [ ]:
X = np.repeat(np.transpose([div_r_gaussian(r_val) for r_val in sorted(r_ej_list)]), 100, axis=0)

In [ ]:
X = np.repeat(np.transpose([div_h_gaussian(h_val) for h_val in sorted(h_ej_list)]), 100, axis=0)

In [ ]:
plt.figure(figsize=(15,15))
plt.imshow(X)

In [ ]:
min(r_ej_list), max(r_ej_list)

In [ ]:
min(h_ej_list), max(h_ej_list)

In [ ]:
plt.plot(sorted(np.clip(np.floor((np.array(r_ej_list)-0.2)/0.05).astype(np.int), 0, 15)))

In [ ]:
Ndat = len(data_pairs_train)
sorted(r_ej_list)